### PR TITLE
feat(@schematics/angular): route generator

### PR DESCRIPTION
**Pull request by hawkgs**
Introduces the lazy routing module generator as part of the existing module generator.

**Feature document:** [Route generator in Angular CLI](https://docs.google.com/document/d/1NAJEWTG54cu4kW7XBp0gVadmz-gqJouW2XDs4c9PPJM/edit?usp=sharing)

### CODE REVIEW

File: packages/schematics/angular/module/files/__name@dasherize@if-flat__/__name@dasherize__.module.ts.template [hunk 1]
```diff
@@ -1,13 +1,20 @@
 import { NgModule } from '@angular/core';<% if (commonModule) { %>
-import { CommonModule } from '@angular/common';<% } %><% if (routing) { %>
-
-import { <%= classify(name) %>RoutingModule } from './<%= dasherize(name) %>-routing.module';<% } %>
+import { CommonModule } from '@angular/common';<% } %><% if (lazyRoute && routeDeclarationInlined) { %>
+import { Routes, RouterModule } from '@angular/router';<% } %>
+<% if (routing || lazyRoute && !routeDeclarationInlined) { %>
```
Comment (alan-agius4) [inline packages/schematics/angular/module/files/__name@dasherize@if-flat__/__name@dasherize__.module.ts.template]: I won't extract these two conditions into template variables that are provided from the schematic itself, as they are bloating the template and they are making it hard to comprehend and view the important bit.

`routing || lazyRoute && !routeDeclarationInlined`
`lazyRoute && routeDeclarationInlined`

File: packages/schematics/angular/utility/project.ts [hunk 2]
```diff
@@ -46,3 +71,19 @@ export function isWorkspaceSchema(workspace: any): workspace is WorkspaceSchema
 export function isWorkspaceProject(project: any): project is WorkspaceProject {
   return !!(project && (project as WorkspaceProject).projectType);
 }
+
+export function isProjectUsingIvy(host: Tree, project: WorkspaceProject): boolean {
+  const projectRoot = getProjectRoot(project);
+  const config = getProjectConfig(host, projectRoot + 'tsconfig.app.json');
```
Comment (alan-agius4) [inline packages/schematics/angular/utility/project.ts]: We need to read this from the workspace, we don't have any guarantee that this wan't changed.

File: packages/schematics/angular/utility/project.ts [hunk 3]
```diff
@@ -35,6 +43,23 @@ export function getProject<TProjectType extends ProjectType = ProjectType.Applic
   return workspace.projects[projectName] as WorkspaceProject<TProjectType>;
 }
 
+export function getProjectConfig(host: Tree, path: string) {
+  const configName = path.split('/').pop();
```
Comment (alan-agius4) [inline packages/schematics/angular/utility/project.ts]: you can use `path.basename(path)` for this.

File: packages/schematics/angular/utility/project.ts [hunk 4]
```diff
@@ -35,6 +43,23 @@ export function getProject<TProjectType extends ProjectType = ProjectType.Applic
...
+  const configFile = host.read(path);
+  if (!configFile) {
+    throw new Error(`Couldn't find project config (${configName}).`);
```
Comment (alan-agius4) [inline packages/schematics/angular/utility/project.ts]: This seem to be related to tsconfig. I would make it clear what config it is that it cannot be found.

File: packages/schematics/angular/utility/project.ts [hunk 5]
```diff
@@ -35,6 +43,23 @@ export function getProject<TProjectType extends ProjectType = ProjectType.Applic
   return workspace.projects[projectName] as WorkspaceProject<TProjectType>;
 }
 
+export function getProjectConfig(host: Tree, path: string) {
```
Comment (alan-agius4) [inline packages/schematics/angular/utility/project.ts]: getProjectTsConfig Maybe? As config is more broad and might imply workspace project config,

File: packages/schematics/angular/module/index.ts [hunk 6]
```diff
@@ -68,6 +70,74 @@ function addDeclarationToNgModule(options: ModuleOptions): Rule {
...
+
+  return path;
+}
+
+function buildRoute(options: ModuleOptions, ivyEnabled: boolean) {
+  let loadChildren: string;
+  const modulePath = `./${options.name}/${options.name}.module`;
+  const moduleName = `${strings.classify(options.name)}Module`;
+
+  if (ivyEnabled) {
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: The import syntax works for VE, hence we should always use this syntax

File: packages/schematics/angular/utility/project.ts [hunk 7]
```diff
@@ -46,3 +71,19 @@ export function isWorkspaceSchema(workspace: any): workspace is WorkspaceSchema
 export function isWorkspaceProject(project: any): project is WorkspaceProject {
   return !!(project && (project as WorkspaceProject).projectType);
 }
+
+export function isProjectUsingIvy(host: Tree, project: WorkspaceProject): boolean {
```
Comment (alan-agius4) [inline packages/schematics/angular/utility/project.ts]: Should be unneeded

File: packages/schematics/angular/utility/project.ts [hunk 8]
```diff
@@ -35,6 +43,23 @@ export function getProject<TProjectType extends ProjectType = ProjectType.Applic
...
+export function getProjectConfig(host: Tree, path: string) {
+  const configName = path.split('/').pop();
+  const configFile = host.read(path);
+  if (!configFile) {
+    throw new Error(`Couldn't find project config (${configName}).`);
+  }
+  const configText = configFile.toString('utf-8');
+  let config: ProjectConfig;
+  try {
+    config = JSON.parse(configText || '{}');
```
Comment (alan-agius4) [inline packages/schematics/angular/utility/project.ts]: TS config can contain comments,
 
You should use the JSON parser.

File: packages/schematics/angular/module/index.ts [hunk 9]
```diff
@@ -68,6 +70,74 @@ function addDeclarationToNgModule(options: ModuleOptions): Rule {
...
+
+    const recorder = host.beginUpdate(path);
+    recorder.insertLeft(addDeclaration.pos, addDeclaration.toAdd);
+    host.commitUpdate(recorder);
+
+    return host;
+  };
+}
+
+function getRoutingModulePath(host: Tree, options: ModuleOptions): Path | undefined {
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: This doesn't seem right. This method will always fall in the `catch` because you cannot locate a routing module with `findModuleFromOptions`

https://github.com/angular/angular-cli/blob/master/packages/schematics/angular/utility/find-module.ts#L110

File: packages/schematics/angular/module/index.ts [hunk 10]
```diff
@@ -68,6 +70,74 @@ function addDeclarationToNgModule(options: ModuleOptions): Rule {
...
+  try {
+    path = findModuleFromOptions(host, { module: routingModuleName, ...rest });
+  } catch {}
+
+  return path;
+}
+
+function buildRoute(options: ModuleOptions, ivyEnabled: boolean) {
+  let loadChildren: string;
+  const modulePath = `./${options.name}/${options.name}.module`;
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: Paths should be dasharized.

File: packages/schematics/angular/module/index.ts [hunk 11]
```diff
@@ -82,18 +152,37 @@ export default function (options: ModuleOptions): Rule {
...
     const templateSource = apply(url('./files'), [
-      options.routing ? noop() : filter(path => !path.endsWith('-routing.module.ts.template')),
+      options.routing || isLazyLoadedModuleGen && !!routingModulePath
+        ? noop()
+        : filter(path => !path.endsWith('-routing.module.ts.template')),
       applyTemplates({
         ...strings,
         'if-flat': (s: string) => options.flat ? '' : s,
+        lazyRoute: isLazyLoadedModuleGen,
+        routeDeclarationInlined: !routingModulePath,
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: Why would we want this? If no routing module is found, we should generate a routing module.

File: packages/schematics/angular/module/index.ts [hunk 12]
```diff
@@ -82,18 +152,37 @@ export default function (options: ModuleOptions): Rule {
...
     ]);
 
     return chain([
-      addDeclarationToNgModule(options),
+      !isLazyLoadedModuleGen ? addDeclarationToNgModule(options) : noop(),
+      addRouteDeclarationToNgModule(options, project, routingModulePath),
+      isLazyLoadedModuleGen
+        ? schematic('component', {
+            ...options,
+            skipImport: true,
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: The lazy loaded component needs to be imported in the lazy loaded module.

File: packages/schematics/angular/module/index.ts [hunk 13]
```diff
@@ -68,6 +70,74 @@ function addDeclarationToNgModule(options: ModuleOptions): Rule {
...
+function addRouteDeclarationToNgModule(
+  options: ModuleOptions,
+  project: WorkspaceProject,
+  routingModulePath: Path | undefined,
+): Rule {
+  return (host: Tree) => {
+    if (!options.route) {
+      return host;
+    }
+    if (options.route && !options.module) {
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: `options.route && !options.module` -> `!options.module`

File: packages/schematics/angular/module/index.ts [hunk 14]
```diff
@@ -68,6 +70,74 @@ function addDeclarationToNgModule(options: ModuleOptions): Rule {
...
+      throw new Error('Module option required when creating a lazy loaded routing module.');
+    }
+
+    let path: string;
+    if (routingModulePath) {
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: The castings here should be redundant otherwise the checks are not correct.

File: packages/schematics/angular/module/index.ts [hunk 15]
```diff
@@ -68,6 +70,74 @@ function addDeclarationToNgModule(options: ModuleOptions): Rule {
...
+      path = options.module as string;
+    }
+
+    const text = host.read(path);
+    if (!text) {
+      throw new Error(`Couldn't find the module nor its routing module.`);
+    }
+
+    const ivyEnabled = isProjectUsingIvy(host, project);
+    const sourceText = text.toString('utf-8');
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: `toString()` should be enough by default it is utf8

File: packages/schematics/angular/module/files/__name@dasherize@if-flat__/__name@dasherize__.module.ts.template (hunk 1 above)
Comment (hawkgs) [reply packages/schematics/angular/module/files/__name@dasherize@if-flat__/__name@dasherize__.module.ts.template]: Do you mean that it is better to stick with using the generator `options` and compose the checks inside the template rather than creating new template variables that incorporate them? Sorry, if I didn't understand you correct.

File: packages/schematics/angular/utility/project.ts (hunk 3 above)
Comment (hawkgs) [reply packages/schematics/angular/utility/project.ts]: Thanks for noting :+1:

File: packages/schematics/angular/utility/project.ts (hunk 4 above)
Comment (hawkgs) [reply packages/schematics/angular/utility/project.ts]: `configName` displays the file name. Am I missing something?

File: packages/schematics/angular/utility/project.ts (hunk 5 above)
Comment (hawkgs) [reply packages/schematics/angular/utility/project.ts]: Yes, totally makes sense. :+1:

File: packages/schematics/angular/utility/project.ts (hunk 8 above)
Comment (hawkgs) [reply packages/schematics/angular/utility/project.ts]: Thanks for the catch!

File: packages/schematics/angular/module/index.ts (hunk 9 above)
Comment (hawkgs) [reply packages/schematics/angular/module/index.ts]: Actually, it should work. Frankly, I haven't looked into the function itself (`findModuleFromOptions`) but if it always used to hit the `catch`, [this test](https://github.com/angular/angular-cli/blob/389f9c7af55063f51d7a78e16af7ab163315346d/packages/schematics/angular/module/index_spec.ts#L148) would always fail.

In other words, we won't be able to access `app-routing.module.ts` and insert the declaration in that case.

File: packages/schematics/angular/module/index.ts (hunk 10 above)
Comment (hawkgs) [reply packages/schematics/angular/module/index.ts]: Thanks for the catch :+1:

File: packages/schematics/angular/module/index.ts (hunk 11 above)
Comment (hawkgs) [reply packages/schematics/angular/module/index.ts]: Because we would like to stick to the preferred by the user strategy, i.e.: route declarations reside in parent module or in parent routing module. If, for example, the parent module has the declarations inlined, we won't generate a child routing module but rather put the routes in the child module. This is part of the document:

> Depending on whether we’ve inserted the route in the parent routing module or in the parent module [i.e. where is the `RouterModule` declared] we’ll determine if we want to generate a child routing module.

File: packages/schematics/angular/module/index.ts (hunk 12 above)
Comment (hawkgs) [reply packages/schematics/angular/module/index.ts]: Ah, I am manually doing this in the template rather than making use of the component schematics functionality. Makes sense.

File: packages/schematics/angular/module/index.ts (hunk 14 above)
Comment (hawkgs) [reply packages/schematics/angular/module/index.ts]: You are right. They are redundant.

File: packages/schematics/angular/module/index.ts (hunk 6 above)
Comment (hawkgs) [reply packages/schematics/angular/module/index.ts]: That's my fault. Fortunately, this will decrease the size of the change this PR introduces.

File: packages/schematics/angular/module/files/__name@dasherize@if-flat__/__name@dasherize__.module.ts.template (hunk 1 above)
Comment (alan-agius4) [reply packages/schematics/angular/module/files/__name@dasherize@if-flat__/__name@dasherize__.module.ts.template]: I meant template options, example
```ts
      applyTemplates({
        ...strings,
        lazyRoute: routing || lazyRoute && !routeDeclarationInlined,
        ...,
      }),
```

File: packages/schematics/angular/module/index.ts (hunk 11 above)
Comment (alan-agius4) [reply packages/schematics/angular/module/index.ts]: Makes sense, thanks

File: packages/schematics/angular/module/index.ts (hunk 12 above)
Comment (alan-agius4) [reply packages/schematics/angular/module/index.ts]: I think this should be done via component schematic.

File: packages/schematics/angular/utility/project.ts (hunk 5 above)
Comment (alan-agius4) [reply packages/schematics/angular/utility/project.ts]: Most likely actually this method will become redundant.

Comment (hawkgs) [reply packages/schematics/angular/utility/project.ts]: Indeed

File: packages/schematics/angular/module/index.ts (hunk 12 above)
Comment (hawkgs) [reply packages/schematics/angular/module/index.ts]: I just pointed that I am doing it the wrong way, apparently. I'll have to keep the import in the routing module though, since we don't need the component added to the declarations array there.

File: packages/schematics/angular/module/files/__name@dasherize@if-flat__/__name@dasherize__.module.ts.template (hunk 1 above)
Comment (hawkgs) [reply packages/schematics/angular/module/files/__name@dasherize@if-flat__/__name@dasherize__.module.ts.template]: I see, but using the `routing` option doesn't necessarily mean that we have a lazy route. Besides, we have both `routing || lazyRoute && !routeDeclarationInlined` && `lazyRoute && routeDeclarationInlined`. 

What about combining just the current `lazyRoute` and `routeDeclarationInlined`? Something like `lazyRouteWithoutRouteMod` and `lazyRouteWithRouteMod`? Still long names though. Do you think that's a good idea?

Comment (alan-agius4) [reply packages/schematics/angular/module/files/__name@dasherize@if-flat__/__name@dasherize__.module.ts.template]: This is good, thought I’d go for the full name, ie: `Module` instead of `Mod` as it’s easier to comprehend.

Comment (hawkgs) [reply packages/schematics/angular/module/files/__name@dasherize@if-flat__/__name@dasherize__.module.ts.template]: Yeah, it's not worth saving 3 letters in the sake of less comprehensive naming. Will do it :+1:

File: packages/schematics/angular/utility/ast-utils.ts [hunk 16]
```diff
@@ -568,3 +579,89 @@ export function isImported(source: ts.SourceFile,
...
+  } else {
+    const routesVarName = routesArg.getText();
+    const routesVar = source.statements
+      .filter((s: ts.Statement) => s.kind === ts.SyntaxKind.VariableStatement)
+      .find((v: ts.VariableStatement) => {
+        return v.declarationList.declarations[0].name.getText() === routesVarName;
+      }) as ts.VariableStatement | undefined;
+
+    if (!routesVar) {
+      throw new Error(`The route declaration variable "${routesVarName}" is used but not defined.`);
```
Comment (clydin) [inline packages/schematics/angular/utility/ast-utils.ts]: Can the error provide file and position information as well?  Same for others where applicable.

Comment (hawkgs) [reply packages/schematics/angular/utility/ast-utils.ts]: I'll see what I can do about that. :+1:

Comment (hawkgs) [reply packages/schematics/angular/utility/ast-utils.ts]: I assume the line of the error and the file name will be enough? Intentionally omitted the column/character position since I think that it may just add an additional noise.

**Comment by alan-agius4** [comment]
Can you please rebase so we make this green?

File: packages/schematics/angular/module/index.ts [hunk 17]
```diff
@@ -68,6 +68,67 @@ function addDeclarationToNgModule(options: ModuleOptions): Rule {
...
+      path = options.module;
+    }
+
+    const text = host.read(path);
+    if (!text) {
+      throw new Error(`Couldn't find the module nor its routing module.`);
+    }
+
+    const sourceText = text.toString();
+    const addDeclaration = addRouteDeclarationToModule(
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: Nit: 
```ts
    const sourceFile = ts.createSourceFile(path, sourceText, ts.ScriptTarget.Latest, true);
    const changes = addRouteDeclarationToModule(source, path, buildRoute(options));
    const output = applyChanges(path, text , changes);
```

**Comment by hawkgs** [comment]
Done.

File: packages/schematics/angular/module/index_spec.ts [hunk 18]
```diff
@@ -118,4 +118,75 @@ describe('Module Schematic', () => {
       .toPromise();
     expect(appTree.files).toContain('/projects/bar/custom/app/foo/foo.module.ts');
   });
+
+  describe('lazy route generator', () => {
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index_spec.ts]: Can we add a couple of tests when the module is flat?

Ie when using the `flat` option?

From the code, I don't see that this is being handled at the moment.

File: packages/schematics/angular/utility/ast-utils.ts [hunk 19]
```diff
@@ -344,6 +344,28 @@ export function getFirstNgModuleName(source: ts.SourceFile): string|undefined {
   return moduleClass.name.text;
 }
 
+export function getMetadataField(
```
Comment (alan-agius4) [inline packages/schematics/angular/utility/ast-utils.ts]: I believe this can be simplified a bit

```ts
export function getMetadataField(
  node: ts.ObjectLiteralExpression,
  metadataField: string,
): ts.ObjectLiteralElement[] {
  return node.properties
    .filter(prop => ts.isPropertyAssignment(prop))
    // Filter out every fields that's not "metadataField". Also handles string literals
    // (but not expressions).
    .filter(({ name }) => {
      return (ts.isIdentifier(name) || ts.isStringLiteral(name))
        && name.text === metadataField;
    });
}


```

Comment (hawkgs) [reply packages/schematics/angular/utility/ast-utils.ts]: Sounds good. By the way, this is more like a question -- is there any valid reason for using `text` property over `getText()`? I noticed that in one of the functions that I introduced, using `text` breaks the tests because it's `undefined`, while `getText` returns `''` (ref: `typescript.js:96792`). Maybe using `getText` is safer since it always returns a string (not that it matters in the current change)?

File: packages/schematics/angular/module/index_spec.ts (hunk 18 above)
Comment (hawkgs) [reply packages/schematics/angular/module/index_spec.ts]: Thanks for the catch! Indeed, I haven't taken care of this. I added a single test since there aren't edge cases that I could come up with.

File: packages/schematics/angular/utility/ast-utils.ts (hunk 19 above)
Comment (alan-agius4) [reply packages/schematics/angular/utility/ast-utils.ts]: When using to compare, returning `undefined` is perfectly fine. using `text` directly. Is slightly more performant.

File: packages/schematics/angular/module/index.ts [hunk 20]
```diff
@@ -68,6 +68,67 @@ function addDeclarationToNgModule(options: ModuleOptions): Rule {
...
+  try {
+    path = findModuleFromOptions(host, { module: routingModuleName, ...rest });
+  } catch {}
+
+  return path;
+}
+
+function buildRoute(options: ModuleOptions) {
+  const dasherizedName = strings.dasherize(options.name);
+  const modulePath = `./${!options.flat ? dasherizedName + '/' : ''}${dasherizedName}.module`;
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: This still doesn't seem right, As a module can be nested when when is flat, while it's flat it is not not in the same folder of the `app.module`

```
ng g m --flat /foo/bar/baz
```

You should re-use the `relativePath` to have the correct module path https://github.com/angular/angular-cli/blob/master/packages/schematics/angular/module/index.ts#L53 instead

Also, you should always take into consideration the `options.path` option.

File: packages/schematics/angular/module/index.ts [hunk 21]
```diff
@@ -82,19 +143,42 @@ export default function (options: ModuleOptions): Rule {
...
+      `${!options.flat ? moduleDasherized + '/' : ''}${moduleDasherized}.module.ts`;
 
     return chain([
-      addDeclarationToNgModule(options),
+      !isLazyLoadedModuleGen ? addDeclarationToNgModule(options) : noop(),
+      addRouteDeclarationToNgModule(options, routingModulePath),
       mergeWith(templateSource),
+      isLazyLoadedModuleGen
+        ? schematic('component', {
+            ...modulelessOptions,
```
Comment (alan-agius4) [inline packages/schematics/angular/module/index.ts]: Nit: 
```
{
...options,
module: modulePath,
}
```

File: packages/schematics/angular/module/index.ts (hunk 20 above)
Comment (hawkgs) [reply packages/schematics/angular/module/index.ts]: Ah, got it.

File: packages/schematics/angular/module/index.ts (hunk 21 above)
Comment (hawkgs) [reply packages/schematics/angular/module/index.ts]: I think it's safer to keep it with `modulelessOptions` because, if we use spread on `options`, we are overriding `module`. That way, the validity of this statement is dependent on the order which might be implicit to the readers (i.e. if someone switches the positions of these two, for some reason, we will end up with a bug). Anyway, that's my personal perception.

**Comment by alan-agius4** [comment]
Thanks for your contribution and the awesome efforts you put in this PR. Keep up the good work, and I hope that we can have you back with another PR soon :)
